Winleafs-Server/Winleafs.Server.Services/SpotifyService.cs

```diff
@@ -0,0 +1,181 @@
+            {
+                { "refresh_token", user.SpotifyRefreshToken },
+                { "grant_type", "refresh_token" },
+            };
+
+            var response = await _spotifyAuthorizationClient.PostAsync("https://accounts.spotify.com/api/token", new FormUrlEncodedContent(postDictionary));
+
+            var result = JsonConvert.DeserializeObject<RefreshTokenResultDTO>(await response.Content.ReadAsStringAsync());
+
+            user.SpotifyAccessToken = result.access_token;
+            user.SpotifyExpiresOn = DateTime.Now.AddSeconds(result.expires_in - 10); //Do minus 10 such that our value is always lower than the expire value on Spotify's side
+
+            _userService.Update(user);
+
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task<string> GetCurrentPlayingPlaylistId(string applicationId)
+        {
+            var webAPI = await GetRefreshedSpotifyWebAPI(applicationId);
+
+            var playbackContext = await webAPI.GetPlaybackAsync();
+
+            if (!playbackContext.IsPlaying)
+            {
+                return null;
+            }
+
+            var playlistUrl = playbackContext.Context.ExternalUrls.FirstOrDefault(externalUrl => externalUrl.Value.Contains("playlist"));
+            return string.IsNullOrEmpty(playlistUrl.Value) ? null : playlistUrl.Value.Remove(0, "https://open.spotify.com/playlist/".Length);
+        }
+
+        public async Task<Dictionary<string, string>> GetPlaylists(string applicationId)
+        {
+            var webAPI = await GetRefreshedSpotifyWebAPI(applicationId);
+            var profile = await webAPI.GetPrivateProfileAsync();
+
+            var info = await webAPI.GetUserPlaylistsAsync(profile.Id);
+
+            return info.Items.ToDictionary(playlist => playlist.Id, playlist => playlist.Name);
+        }
+
+        public async Task Disconnect(string applicationId)
+        {
+            var user = await _userService.FindUserByApplicationId(applicationId);
+
+            if (user != null)
+            {
+                user.SpotifyAccessToken = null;
+                user.SpotifyExpiresOn = null;
+                user.SpotifyRefreshToken = null;
+
+                _userService.Update(user);
+
+                await _context.SaveChangesAsync();
+            }
+        }
+
+        public async Task<bool> IsConnected(string applicationId)
+        {
+            try
+            {
+                var webAPI = await GetRefreshedSpotifyWebAPI(applicationId);
+
+                var playbackContext = await webAPI.GetPlaybackAsync();
+
+                return playbackContext.Error == null;
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
+        #region Privates
+        private string GetSpotifyAuthorizationHeader()
+        {
+            var spotifyauthorizationBytes = Encoding.UTF8.GetBytes($"{SpotifyClientInfo.ClientID}:{SpotifyClientInfo.ClientSecret}");
```

Comment:
using

Reply:
What would using improve here?

Reply:
Nvm I don't know how the encoding library works apparently.